src/software/simulated_tests/simulated_test_fixture.cpp

```diff
@@ -174,31 +178,33 @@ void SimulatedTestFixture::runTest(
                                                         CAMERA_FRAMES_PER_AI_TICK);
 
     // Tick one frame to aid with visualization
-    bool validation_functions_done =
-        tickTest(terminating_validation_functions, non_terminating_validation_functions,
-                 simulation_time_step, ai_time_step, world);
+    bool validation_functions_done = tickTest(simulation_time_step, ai_time_step, world);
     while (simulator->getTimestamp() < timeout_time && !validation_functions_done)
     {
         if (!thunderbots_config->getAiControlConfig()->getRunAi()->value())
         {
+            auto ms_to_sleep = std::chrono::milliseconds(
+                static_cast<int>(ai_time_step.toMilliseconds()));
+            std::this_thread::sleep_for(ms_to_sleep);
             continue;
         }
-        validation_functions_done = tickTest(terminating_validation_functions,
-                                             non_terminating_validation_functions,
-                                             simulation_time_step, ai_time_step, world);
+        validation_functions_done = tickTest(simulation_time_step, ai_time_step, world);
     }
 
     if (!validation_functions_done && !terminating_validation_functions.empty())
     {
-        ADD_FAILURE()
-            << "Not all validation functions passed within the timeout duration";
+        std::string failure_message =
+            "Not all validation functions passed within the timeout duration:\n";
+        for (const auto &fun : terminating_function_validators)
+        {
+            failure_message += fun.currentErrorMessage() + std::string("\n");
```

Comment:
I couldn't fully track it down, but is it possible for `fun.currentErrorMessage()` to be `null` or `""` if there are no error messages? As that could fill up the `failure_message` with a bunch of empty lines.

Reply:
good point, added a check for that